### PR TITLE
feat: setup invitation, reminder, and expiry emails for payment requests using Hasura event triggers

### PR DESCRIPTION
**Pull request by jessicamcinchak**
when a `payment_requests` record is inserted (do each x2 for payee & agent):
- immediately send an invitation to pay email
- schedule a reminder email in 28-7 days
- schedule an expiry email in 28 days

todo: 
- [x] update s3 secrets with new Notify template ids
- [x] populate basic email body templates with session preview data, team notify personalisation, etc - share with Ian/design team for feedback
- [x] test coverage
- [ ] clean up scheduled reminders & expiry events when the `payment_request` is marked as paid - would vote this is a done in a follow-up PR! 

to test: 
- generate a lowcal_session record, copy the sessionId and manually insert a related `payment_requests` record via the Hasura console (or "clone" an existing row, swapping in your own email)
- click "save" to insert: you should get 2 emails in your inbox (assumes you've used your email for both the session & payment)
- you should be able to check the Hasura "events" tab and see successfully scheduled events for reminder & expiry emails

### CODE REVIEW

**Comment by jessicamcinchak** [comment]
[[Hasura events] Send invitation, reminder, and expiry emails to payee and agent](https://trello.com/c/UQhzSydw/2354-hasura-events-send-invitation-reminder-and-expiry-emails-to-payee-and-agent)

**Comment by jessicamcinchak** [comment]
[[Notify] Create templates for invitation, reminder, expiry (x2 one each for payee & agent)](https://trello.com/c/poV15sP2/2351-notify-create-templates-for-invitation-reminder-expiry-x2-one-each-for-payee-agent)

File: api.planx.uk/inviteToPay/sendPaymentEmail.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        paymentRequestId: paymentRequestId,
+        payeeEmail: payment.payee_email,
+        payeeName: payment.payee_name,
+        agentName: `{TODO}`,
+        paymentLink: `https://${payment.session.flow.team.domain}/${payment.session.flow.slug}/pay?paymentRequestId=${paymentRequestId}`,
+        fee: (payment.session_preview_data.fee).toLocaleString("en-GB", { style: "currency", currency: "GBP" }),
+        projectType: payment.session_preview_data?.["proposal.projectType"],
+        address: payment.session_preview_data?._address.title,
+        expiryDate: calculateExpiryDate(payment.created_at),
+      }
```
Comment (jessicamcinchak) [inline api.planx.uk/inviteToPay/sendPaymentEmail.ts:30]: this `personalisation` type & mapping is my first draft / best guess - I fully anticipate we'll keep refining this as we refine the `payment_requests` schema and the content team take a pass at the Notify templates themselves. 

it should be simple to pick up in small, follow-up PRs.

File: api.planx.uk/saveAndReturn/sendEmail.ts [hunk 2]
```diff
@@ -1,25 +1,35 @@
 import { NextFunction, Request, Response } from "express";
 import { sendSingleApplicationEmail, Template } from "./utils";
+import { sendSinglePaymentEmail } from "../inviteToPay";
 
 const sendSaveAndReturnEmail = async (
```
Comment (jessicamcinchak) [inline api.planx.uk/saveAndReturn/sendEmail.ts:5]: It made sense to me to use the existing `/send-email/:template` endpoint defined in server.js, even though that calls this function - whose name has now fallen out of sync with what it does. But I failed to think of something better - any ideas? 

`routeEmailTemplate` maybe? It's not actually responsible for sending the email (and never has been), but rather it checks the provided payload and determines if we should then call a method to validate/send a Save&Return email or an Invite to Pay one.

File: api.planx.uk/saveAndReturn/utils.ts [hunk 3]
```diff
@@ -56,7 +63,7 @@ export type Template = keyof typeof emailTemplates;
 const sendEmail = async (
   template: Template,
   emailAddress: string,
-  config: SaveAndReturnNotifyConfig | EmailSubmissionNotifyConfig
+  config: SaveAndReturnNotifyConfig | EmailSubmissionNotifyConfig | InviteToPayNotifyConfig
```
Comment (jessicamcinchak) [inline api.planx.uk/saveAndReturn/utils.ts:66]: it'd be nice to think more intentionally about these types and what can be shared going forward: eg is there a `MinimumNotifyPersonalisation` type that we could then extend for domain-specific templates ?

i took a quick pass at this and introduced some unexpected errors, so just settled on introducing a discrete new type here to keep the scope of changes in check. but one to come back to soon!

File: api.planx.uk/webhooks/paymentRequestEvents.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
+import { addDays } from 'date-fns';
+import { Request, Response, NextFunction } from 'express';
+
+import { createScheduledEvent } from '../hasura/metadata';
+import { DAYS_UNTIL_EXPIRY } from '../saveAndReturn/utils';
```
Comment (jessicamcinchak) [inline api.planx.uk/webhooks/paymentRequestEvents.ts:5]: worth pointing out that this aligns the payment request "validity period" to `lowcal_sessions` - both at 28 days. may make sense to treat uniquely in the future, but this felt straightforward for now until we have guidance that suggests another time period

File: api.planx.uk/inviteToPay/sendPaymentEmail.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        emailReplyToId: payment.session.flow.team.notify_personalisation.emailReplyToId,
+        helpEmail: payment.session.flow.team.notify_personalisation.helpEmail,
+        helpOpeningHours: payment.session.flow.team.notify_personalisation.helpOpeningHours,
+        helpPhone: payment.session.flow.team.notify_personalisation.helpPhone,
+        paymentRequestId: paymentRequestId,
+        payeeEmail: payment.payee_email,
+        payeeName: payment.payee_name,
+        agentName: `{TODO}`,
+        paymentLink: `https://${payment.session.flow.team.domain}/${payment.session.flow.slug}/pay?paymentRequestId=${paymentRequestId}`,
+        fee: (payment.session_preview_data.fee).toLocaleString("en-GB", { style: "currency", currency: "GBP" }),
```
[the remark this answers is not in the document]
Comment (DafyddLlyr) [reply api.planx.uk/inviteToPay/sendPaymentEmail.ts:26]: The code here looks correct, but I'm just getting a plain number in the agent email - 
image

**Review by DafyddLlyr — APPROVED** [review]
A few small nits - as always, please feel free to ignore whatever you see fit! The only thing that we really need to address I think is the fee currency issue.

Looking really good though - nice and clear, and a good level of tests it seems.

Works as expected - 6 emails triggered when a payment request is inserted 👍

File: api.planx.uk/types.ts [hunk 6]
```diff
@@ -90,6 +90,26 @@ export interface EmailSubmissionNotifyConfig {
...
+    paymentRequestId: string;
+    payeeEmail: string;
+    payeeName?: string;
+    agentName?: string;
+    paymentLink?: string;
+    fee?: string;
+    projectType?: string;
+    address?: any;
+    expiryDate?: string;
+    id: string; // sessionId
```
[the remark this answers is not in the document]
Comment (jessicamcinchak) [reply api.planx.uk/types.ts:104]: this ties into the point above about the `sendEmail` "config" type which annoying requires `id` to exist across all configs - hence this naming. will add to the followup tech debt card so we can address/clarify the naming all in one swoop!

File: api.planx.uk/inviteToPay/sendPaymentEmail.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jessicamcinchak) [reply api.planx.uk/inviteToPay/sendPaymentEmail.ts:26]: ah good shout, I'm getting the same - looking into how Notify is treating this now / maybe it doesn't like the currency symbol?

File: api.planx.uk/inviteToPay/sendPaymentEmail.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        projectType: payment.session_preview_data?.["proposal.projectType"],
```
[the remark this answers is not in the document]
Comment (jessicamcinchak) [reply api.planx.uk/inviteToPay/sendPaymentEmail.ts:27]: Yes good shout - I thought about this and then I think I noticed a sample `session_preview_data` json a bit back that had pre-cleaned values so I didn't come back to this, but I agree it makes sense to store raw passport values and for the responsibility of formatting to live here :+1:

File: .env.example [hunk 8]
```diff
@@ -68,13 +68,25 @@ GOV_UK_PAY_TOKEN_LAMBETH=👻
...
+
+# Save & return templates
 GOVUK_NOTIFY_SAVE_RETURN_EMAIL_TEMPLATE_ID=👻
 GOVUK_NOTIFY_RESUME_EMAIL_TEMPLATE_ID=👻
 GOVUK_NOTIFY_REMINDER_EMAIL_TEMPLATE_ID=👻
 GOVUK_NOTIFY_EXPIRY_EMAIL_TEMPLATE_ID=👻
-GOVUK_NOTIFY_SUBMISSION_EMAIL_TEMPLATE_ID=👻
 GOVUK_NOTIFY_CONFIRMATION_EMAIL_TEMPLATE_ID=👻
 
+# Invite to pay templates
```
[the remark this answers is not in the document]
Comment (jessicamcinchak) [reply .env.example:79]: Yep agree these aren't actually "secrets" and that 6 new templates felt like a lot of unnecessary length to add to .env & docker configs - I think this one will be simplest to tackle as part of the followup Notify tech debt card here https://trello.com/c/b1HtcRgM/2371-improve-api-handling-for-configuring-and-sending-govuk-notify-emails

File: api.planx.uk/types.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (DafyddLlyr) [reply api.planx.uk/types.ts:104]: Ah perfect - thanks for the explanation - figured there would be a reason but couldn't spot it.

File: api.planx.uk/inviteToPay/sendPaymentEmail.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (DafyddLlyr) [reply api.planx.uk/inviteToPay/sendPaymentEmail.ts:26]: I think I got it - it's an issue with using an integer and not a number. TIL!

image

File: api.planx.uk/saveAndReturn/sendEmail.ts (hunk 2 above)
Comment (jessicamcinchak) [reply api.planx.uk/saveAndReturn/sendEmail.ts]: updated throughout to be `routeSendEmailRequest` :+1:

still lives within `saveAndReturn` subdirectory, but added file re-organisation to Trello debt ticket :broom:

File: api.planx.uk/inviteToPay/sendPaymentEmail.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jessicamcinchak) [reply api.planx.uk/inviteToPay/sendPaymentEmail.ts:27]: okay, I've remembered why I didn't format project type here yet - `getHumanReadableProjectType()` takes a whole LowcalSession as an argument, whereas all I want to pass here is an array of passport variables from `session_preview_data` (we're purposely trying _not_ to pass all the session data around at the payment stage).

Are you okay with this merging as unformatted for now and I'll address in a followup refactor/PR?

[the remark this answers is not in the document]
Comment (DafyddLlyr) [reply api.planx.uk/inviteToPay/sendPaymentEmail.ts:27]: Yep - makes perfect sense 👍

**Pull request merged**